src/infra/core/tests/tests/engine/test_engine_ingest.rs

```diff
@@ -85,87 +100,816 @@ async fn test_ingest_csv_with_engine() {
+                            .to_string(),
+                        },
+                        SqlQueryStep {
+                            alias: None,
+                            query: indoc::indoc!(
+                                r#"
+                                select
+                                    city,
+                                    population + 1 as population
+                                from step1
+                                "#
+                            )
+                            .to_string(),
+                        },
+                    ]),
+                    temporal_tables: None,
+                })
+                .merge(MergeStrategySnapshot {
+                    primary_key: vec!["city".to_string()],
+                    compare_columns: None,
+                    observation_column: None,
+                    obsv_added: None,
+                    obsv_changed: None,
+                    obsv_removed: None,
+                })
+                .build(),
+        )
+        .build();
+
+    let dataset_name = dataset_snapshot.name.clone();
+
+    harness.create_dataset(dataset_snapshot).await;
+
+    // Round 1
+    std::fs::write(
+        &src_path,
+        indoc!(
+            "
+            city,population
+            A,1000
+            B,2000
+            C,3000
+            "
+        ),
     )
     .unwrap();
 
-    let arrow_writer_props = WriterProperties::builder()
-        .set_compression(Compression::GZIP(GzipLevel::default()))
-        .build();
+    harness.ingest(&dataset_name).await.unwrap();
+
+    let df = harness.get_last_data(&dataset_name).await;
+    kamu_data_utils::testing::assert_schema_eq(
+        df.schema(),
+        indoc!(
+            r#"
+            message arrow_schema {
+              REQUIRED INT64 offset;
+              REQUIRED INT64 system_time (TIMESTAMP(MILLIS,true));
+              REQUIRED INT64 event_time (TIMESTAMP(MILLIS,true));
+              REQUIRED BYTE_ARRAY observed (STRING);
+              OPTIONAL BYTE_ARRAY city (STRING);
+              OPTIONAL INT64 population;
+            }
+            "#
+        ),
+    );
+
+    kamu_data_utils::testing::assert_data_eq(
+        df,
+        indoc!(
+            r#"
+            +--------+----------------------+----------------------+----------+------+------------+
+            | offset | system_time          | event_time           | observed | city | population |
+            +--------+----------------------+----------------------+----------+------+------------+
+            | 0      | 2050-01-01T12:00:00Z | 2050-01-01T12:00:00Z | I        | A    | 10001      |
+            | 1      | 2050-01-01T12:00:00Z | 2050-01-01T12:00:00Z | I        | B    | 20001      |
+            | 2      | 2050-01-01T12:00:00Z | 2050-01-01T12:00:00Z | I        | C    | 30001      |
+            +--------+----------------------+----------------------+----------+------+------------+
+            "#
```

Comment:
Nice. I think I had a test somewhere where I did the `DataFrame` unpacking manually to compare 1 value.

Reply:
Thanks, I tried a dozen different approaches to writing tests for `DataFrame`s and this seems like the least verbose and flexible one I had.

Reply:
I'm also using [pretty_assertions crate](https://docs.rs/pretty_assertions/latest/pretty_assertions/) making errors very easy to read.